### PR TITLE
fix: refactor query logic

### PR DESCRIPTION
**Pull request by achingbrain**
1. Queries are plain functions now
1. Queries are run by the QueryManager
1. Query functions yield events - dial, send request, receive response, value, error, etc
1. Timeouts have been replaced with AbortSignals
1. Halting in-progress queries is done by aborting an AbortSignal from the top level or breaking out of the `for await..of` loop
1. Internal components do not depend on the top-level DHT component any more
1. Some methods have been moved to other components to break circular dependencies
1. The top level KadDHT class now has a ts interface and a factory function
1. There are two DHTs, lan (for private addresses) and wan (for public addresses)
1. Client mode is now on by default for the wan DHT, off for the lan DHT
1. DHT queries for CIDs search for the multihash rather than the full CID

### CODE REVIEW

File: src/query-manager/disjoint-path.js [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  cleanUp.on('cleanup', () => {
+    cleanup()
+    deferred.resolve()
+  })
+
+  while (running) { // eslint-disable-line no-unmodified-loop-condition
+    // TODO: ensure we can break out of this loop early - this promise needs to
+    // be resolved manually and will leak memory otherwise
+    await deferred.promise
+    deferred = defer()
```
Comment (kumavis) [inline src/query-manager/disjoint-path.js:222]: maybe can be replaced by a for-await ?

File: src/query-manager/disjoint-path.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+    // yield all available results
+    while (results.length) {
```
Comment (kumavis) [inline src/query-manager/disjoint-path.js:225]: can you do a yield-star here?

File: src/query-manager/disjoint-path.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        // if there are closer peers and the query has not completed, continue the query
+        if (result && !result.done && result.closerPeers) {
+          for (const closerPeer of result.closerPeers) {
+            const closerPeerKadId = await convertPeerId(closerPeer)
+            const closerPeerXor = BigInt('0x' + toString(xor(closerPeerKadId, kadId), 'base16'))
+
+            // only continue query if closer peer is actually closer
+            if (closerPeerXor > peerXor) {
+              // log('skipping %p as they are not closer to %b than %p', closerPeer, key, peer)
+              // continue
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/query-manager/disjoint-path.js:99]: It causes [this test](https://github.com/libp2p/js-libp2p-kad-dht/blob/fix/refactor-query-logic/test/query-manager.spec.js#L851-L911) to pass but then [this test](https://github.com/libp2p/js-libp2p-kad-dht/blob/fix/refactor-query-logic/test/kad-dht.spec.js#L580-L597) to fail.

There's a bug in there somewhere, I wanted to experiment a bit more with go-ipfs and the [expanded interop test suite](https://github.com/ipfs/interop/pull/376) before returning to this.

**Comment by jacobheun** [comment]
Nice! Will try to carve out some time today to go through this. If you haven't already gone through and checked the implementation summary doc, https://github.com/libp2p/js-libp2p-kad-dht/blob/fix/refactor-query-logic/docs/IMPL_SUMMARY.MD, it would be helpful to make sure that's up to date, especially if we're doing anything different/in addition to the dht spec.

**Comment by achingbrain** [comment]
@jacobheun I don't think we're doing anything extra to the spec, at least, we shouldn't be.

That document is interesting, it seems to be mostly covered by https://docs.ipfs.io/concepts/dht/ - perhaps we could link to that instead?

**Comment by achingbrain** [comment]
Ha, well ok then in that case we are doing extra to the spec as we're using disjoint paths.  Probably a good thing though.

**Review by jacobheun — APPROVED** [review]
Some minor things, overall looks good, great improvements to the query logic. @achingbrain and I reviewed on a call to walk through the changes.

File: src/constants.js [hunk 4]
```diff
@@ -33,3 +33,5 @@ exports.K = 20
 
 // Alpha is the concurrency for asynchronous requests
 exports.ALPHA = 3
+
+exports.QUERY_SELF_INTERVAL = Number(minute)
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/constants.js:37]: It's exposed as the `querySelfInterval` factory function option.  Will increase the default to 5 mins

File: src/index.js [hunk 5]
```diff
@@ -1,293 +1,208 @@
...
-    this.providers = new Providers(this.datastore, this.peerId)
-
-    this.validators = {
-      pk: libp2pRecord.validator.validators.pk,
-      ...validators
-    }
+    this._wan = wan
+    this._lan = lan
+    this._libp2p = libp2p
+    this._datastore = libp2p.datastore || this._wan._datastore
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/index.js:60]: We only store `libp2p-record`s in the datastore and they are externally validated so I don't think so, no.

File: src/index.js [hunk 6]
```diff
@@ -1,293 +1,208 @@
...
 
-    // datastore events
-    this.onPut = onPut
-    this.onRemove = onRemove
+  /**
+   * Whether we are in client or server mode
+   */
+  enableServerMode () {
+    log('enabling server mode')
+    this._wan.enableServerMode()
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/index.js:92]: lan is server-mode only

**Pull request merged**